examples/unix/c11/z_bytes.c

```diff
@@ -22,180 +22,199 @@
+        z_bytes_reader_deserialize_sequence_begin(&reader, &num_elements);
+        assert(num_elements == 2);
+        kv_pair_t kvs_output[2];
+        for (size_t i = 0; i < num_elements; ++i) {
+            z_bytes_reader_deserialize_int32(&reader, &kvs_output[i].key);
+            z_bytes_reader_deserialize_string(&reader, &kvs_output[i].value);
+        }
+
+        for (size_t i = 0; i < 2; ++i) {
+            assert(kvs_input[i].key == kvs_output[i].key);
+            assert(strncmp(z_string_data(z_loan(kvs_input[i].value)), z_string_data(z_loan(kvs_output[i].value)),
+                           z_string_len(z_loan(kvs_input[i].value))) == 0);
+            z_drop(z_move(kvs_input[i].value));
+            z_drop(z_move(kvs_output[i].value));
+        }
+        z_drop(z_move(payload));
     }
-}
 
-static void parse_hashmap(kv_pairs_rx_t *kvp, const z_loaned_bytes_t *hashmap) {
-    z_owned_bytes_t kv, first, second;
-    z_bytes_iterator_t iter = z_bytes_get_iterator(hashmap);
-
-    while (kvp->current_idx < kvp->len && z_bytes_iterator_next(&iter, &kv)) {
-        z_bytes_deserialize_into_pair(z_loan(kv), &first, &second);
-        z_bytes_deserialize_into_string(z_loan(first), &kvp->data[kvp->current_idx].key);
-        z_bytes_deserialize_into_string(z_loan(second), &kvp->data[kvp->current_idx].value);
-        z_bytes_drop(z_bytes_move(&first));
-        z_bytes_drop(z_bytes_move(&second));
-        z_bytes_drop(z_bytes_move(&kv));
-        kvp->current_idx++;
+    {
+        // custom struct/tuple serializaiton
+        custom_struct_t cs = (custom_struct_t){.f = 1.0f, .u = {{1, 2, 3}, {4, 5, 6}}, .c = "test"};
+
+        z_bytes_writer_t writer = z_bytes_get_writer(z_loan_mut(payload));
+        z_bytes_writer_serialize_float(&writer, cs.f);
+        z_bytes_writer_serialize_sequence_begin(&writer, 2);
+        for (size_t i = 0; i < 2; ++i) {
+            z_bytes_writer_serialize_sequence_begin(&writer, 3);
+            for (size_t j = 0; j < 3; ++j) {
+                z_bytes_writer_serialize_uint64(&writer, cs.u[i][j]);
+            }
+            z_bytes_writer_serialize_sequence_end(&writer);
+        }
+        z_bytes_writer_serialize_sequence_end(&writer);
+        z_bytes_writer_serialize_str(&writer, cs.c);
+
+        float f = 0.0f;
+        uint64_t u = 0;
+        z_owned_string_t c;
+
+        z_bytes_reader_t reader = z_bytes_get_reader(z_loan(payload));
+        z_bytes_reader_deserialize_float(&reader, &f);
+        assert(f == cs.f);
+        size_t num_elements0 = 0;
+        z_bytes_reader_deserialize_sequence_begin(&reader, &num_elements0);
+        assert(num_elements0 == 2);
+        for (size_t i = 0; i < 2; ++i) {
+            size_t num_elements1 = 0;
+            z_bytes_reader_deserialize_sequence_begin(&reader, &num_elements1);
+            assert(num_elements1 == 3);
+            for (size_t j = 0; j < 3; ++j) {
+                z_bytes_reader_deserialize_uint64(&reader, &u);
```

Comment:
deserialize function calls will return a error if we run out of bytes to read, otherwise there are no any checks so it is up to the user to ensure he serializes/deserializes data correctly.

Reply:
We definitely need to provide at least some protection or checks, but this can be done later, the main thing is that we have a returned error code, which we can then expand.